### PR TITLE
ENG-22805: ensure correct validation of server certificate

### PR DESCRIPTION
**Pull request by rdykiel**
Pleas let me know what tests I should be running, and also how I should modify mod.go prior to merging

### CODE REVIEW

File: examples/tls/tls_connect.go [hunk 1]
```diff
@@ -0,0 +1,106 @@
+/* This file is part of VoltDB.
+ * Copyright (C) 2008-2018 VoltDB Inc.
```
Comment (dave-voltdb) [inline examples/tls/tls_connect.go:2]: company name
date

File: go.mod [hunk 2]
```diff
@@ -1,9 +1,12 @@
-module github.com/VoltDB/voltdb-client-go
+module voltdb-client-go
 
 go 1.16
 
+replace github.com/VoltDB/voltdb-client-go => ../voltdb-client-go
```
Comment (akhanzode) [inline go.mod:5]: remove replace this is for your local testing

Comment (akhanzode) [reply go.mod:5]: Do you need to modify go mod and sum? are you bringing new packages?

File: go.mod [hunk 3]
```diff
@@ -1,9 +1,12 @@
-module github.com/VoltDB/voltdb-client-go
+module voltdb-client-go
```
Comment (dave-voltdb) [inline go.mod:1]: I don't understand the use of this file so cannot review

File: voltdbclient/node_conn.go [hunk 4]
```diff
@@ -235,19 +234,26 @@ func (nc *nodeConn) networkConnect(protocolVersion int) (interface{}, *wire.Conn
 			if !ok {
 				log.Fatal("failed to parse root certificate")
 			}
-			// Use PEM file as your tlsconfig
-			nc.tlsConfig = &tls.Config{
-				RootCAs:            roots,
-				InsecureSkipVerify: nc.insecureSkipVerify,
-			}
+
+			// Set up a config using PERM contents as the root CAs
```
Comment (dave-voltdb) [inline voltdbclient/node_conn.go:238]: PEM

File: examples/tls/tls_connect.go [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+			InsecureSkipVerify: insecure,
+			ServerName:         altServer,
+		}
+	}
+
+	conn, err := voltdbclient.OpenTLSConn(server, voltdbclient.ClientConfig{
+		PEMPath:            pemPath,
+		TLSConfig:          tlsConfig,
+		InsecureSkipVerify: insecure,
+		ConnectTimeout:     60000000,
```
Comment (akhanzode) [inline examples/tls/tls_connect.go:72]: You should configure and pass down timeout.

**Review by dave-voltdb — COMMENTED** [review]
Looks plausible, but ANish needs to look at it.

File: examples/tls/tls_connect.go (hunk 5 above)
Comment (akhanzode) [reply examples/tls/tls_connect.go:72]: never mind its example code.

File: voltdbclient/node_conn.go [hunk 6]
```diff
@@ -235,19 +234,26 @@ func (nc *nodeConn) networkConnect(protocolVersion int) (interface{}, *wire.Conn
...
+			tlsConfigCopy.RootCAs = roots
+			nc.tlsConfig = tlsConfigCopy
 		}
 		dialer := net.Dialer{
 			Timeout: to,
 		}
+
+		// In secure mode, go requires a ServerName, so force it if absent
+		if nc.tlsConfig.ServerName == "" && !nc.tlsConfig.InsecureSkipVerify {
+			nc.tlsConfig.ServerName = nc.connInfo
```
Comment (akhanzode) [inline voltdbclient/node_conn.go:249]: Check for nil?

File: examples/tls/tls_connect.go (hunk 1 above)
Comment (rdykiel) [reply examples/tls/tls_connect.go:2]: Fixed

File: go.mod (hunk 3 above)
Comment (rdykiel) [reply go.mod:1]: Will revert all changes to these

File: voltdbclient/node_conn.go (hunk 4 above)
Comment (rdykiel) [reply voltdbclient/node_conn.go:238]: Fixed

File: go.mod (hunk 2 above)
Comment (rdykiel) [reply go.mod:5]: Reverting the changes to go.mod and go.sum

File: voltdbclient/node_conn.go (hunk 6 above)
Comment (rdykiel) [reply voltdbclient/node_conn.go:249]: Added

**Review by akhanzode — APPROVED** [review]

**Pull request merged**